stdlib/LinearAlgebra/src/matmul.jl

```diff
@@ -182,6 +182,53 @@ for elty in (Float32,Float64)
     end
 end
 
+"""
+    muladd(A, y, z)
+
+Combined multiply-add, `A*y .+ z`, for matrix-matrix or matrix-vector multiplication.
+The result is always the size of `A*y`, but `z` may be smaller, or a scalar.
+
+!!! compat "Julia 1.6"
+     These methods require Julia 1.6 or later.
+
+# Examples
+```jldoctest
+julia> A=[1.0 2.0; 3.0 4.0]; B=[1.0 1.0; 1.0 1.0]; C=[0, 100];
+
+julia> muladd(A, B, C)
+2×2 Matrix{Float64}:
+   3.0    3.0
+ 107.0  107.0
+```
+"""
+function Base.muladd(A::AbstractMatrix{TA}, y::AbstractVector{Ty}, z) where {TA, Ty}
+    T = promote_type(TA, Ty, eltype(z))
+    C = similar(A, T, axes(A,1))
+    C .= z
+    mul!(C, A, y, true, true)
+end
+
+function Base.muladd(A::AbstractMatrix{TA}, B::AbstractMatrix{TB}, z) where {TA, TB}
+    T = promote_type(TA, TB, eltype(z))
+    C = similar(A, T, axes(A,1), axes(B,2))
+    C .= z
+    mul!(C, A, B, true, true)
+end
+
+Base.muladd(x::AdjointAbsVec, A::AbstractMatrix, z) = muladd(A', x', z')'
+Base.muladd(x::TransposeAbsVec, A::AbstractMatrix, z) = transpose(muladd(transpose(A), transpose(x), transpose(z)))
+
+function Base.muladd(u::AbstractVector, v::AdjOrTransAbsVec, z)
+    ndims(z) > 2 && throw(DimensionMismatch("cannot broadcast array to have fewer dimensions"))
+    (u .* v) .+ z
+end
+
+function Base.muladd(u::AdjOrTransAbsVec, v::AbstractVector, z)
+    uv = _dot_nonrecursive(u, v)
+    ndims(z) > ndims(uv) && throw(DimensionMismatch("cannot broadcast array to have fewer dimensions"))
```

Comment:
Is there any way one could have this test before computing `_dot_nonrecursive`?

Reply:
This is the only multiplication which produces a scalar (on ordinary arrays) and would then expect z nothing more than a scalar. 

On arrays of arrays, it may produce one sub-array, and in this case `z` is allowed to be the same as that. One could try to infer whether this will happen before  `_dot_nonrecursive`, but what would we gain from the extra complexity?

Reply:
I just thought that one may wish to throw before doing "heavy" computations, and not first work hard just to eventually realize that it wouldn't work.

If it's a mess to figure that out beforehand, I think it's fine.

Reply:
Worst case would be something like `Any[ones(2,1) for _ in 1:3]' * Any[ones(2) for _ in 1:3] .+ [1]` where you can't tell from the types that it's safe. 

I guess an inner product will seldom be the biggest step in the your calculation, too, so hopefully not a big penalty to wait for the error.